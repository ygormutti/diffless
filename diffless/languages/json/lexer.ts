import jsonTokenize = require('json-tokenize');
import { Token, ValuedToken } from '../../hcsdiff/model';
import { Document, DocumentURI, Location, Position, Range } from '../../model';

enum TokenType {
    LeftBrace = 'LeftBrace',
    RightBrace = 'RightBrace',
    Semicolon = 'Semicolon',
    Comma = 'Comma',
    LeftBracket = 'LeftBracket',
    RightBracket = 'RightBracket',
    String = 'String',
    Number = 'Number',
    Boolean = 'Boolean',
    Null = 'Null',
}

export function tokenize(document: Document): Token[] {
    return jsonTokenize(document.content)
        .map(token => adapt(document.uri, token))
        .filter(token => token !== undefined) as Token[];
}

export function adapt(uri: DocumentURI, token: JsonTokenize.Token): Token | undefined {
    const { type, value } = token;
    switch (type) {
        case 'literal':
            switch (value) {
                case null:
                    return toToken(uri, token, TokenType.Null);
                case true:
                case false:
                    return toValuedToken(uri, token, TokenType.Boolean);
            }
            break;

        case 'number':
            return toValuedToken(uri, token, TokenType.Number);

        case 'punctuation':
            switch (value) {
                case '{':
                    return toToken(uri, token, TokenType.LeftBrace);
                case '}':
                    return toToken(uri, token, TokenType.RightBrace);
                case '[':
                    return toToken(uri, token, TokenType.LeftBracket);
                case ']':
                    return toToken(uri, token, TokenType.RightBracket);
                case ':':
                    return toToken(uri, token, TokenType.Semicolon);
                case ',':
                    return toToken(uri, token, TokenType.Comma);
            }

        case 'string':
            return toValuedToken(uri, token, TokenType.String);

        case 'whitespace':
            return undefined;
    }
}

function toToken(uri: DocumentURI, token: JsonTokenize.Token, type: TokenType) {
    const { raw, position } = token;
    return new Token(toLocation(uri, raw, position), raw, type);
}

function toValuedToken(uri: DocumentURI, token: JsonTokenize.Token, type: TokenType) {
    const { raw, position } = token;
    return new ValuedToken(toLocation(uri, raw, position), raw, type, token.value, (a, b) => a === b);
}

function toLocation(uri: DocumentURI, raw: string, locator: JsonTokenize.Range | JsonTokenize.Position) {
    return new Location(uri, toRange(raw, locator));
}

function toRange(raw: string, locator: JsonTokenize.Range | JsonTokenize.Position) {
    let start;
    let end;
    if (raw.length === 1) {
        const position = locator as JsonTokenize.Position;
        start = new Position(position.lineno, position.column);
        end = new Position(position.lineno, position.column + 1);
    } else {
        const range = locator as JsonTokenize.Range;
        start = new Position(range.start.lineno, range.start.column);
        end = new Position(range.end.lineno, range.end.column + 1);
    }
    return new Range(start, end);
}
