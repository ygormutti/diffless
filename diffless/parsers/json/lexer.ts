enum TokenTypes {
    LeftBrace,
    RightBrace,
    Semicolon,
    Comma,
    LeftBracket,
    RightBracket,
    String,
    Number,
    True,
    False,
    Null
}

function lex(source: string) {
    for (let token of tokenizer(source)) {
        switch (token.type) {
            case tokenizer.TokenType.Punctuation:
                switch (token.raw) {
                    case '{':

                        break;
                
                    default:
                        break;
                }
                break;
        
            default:
                break;
        }
    }
}