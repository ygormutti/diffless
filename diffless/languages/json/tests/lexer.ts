import { fixtureDocument } from '../../../tests/test-util';
import { tokenize } from '../lexer';

describe('lexer', () => {
    describe('tokenize', () => {
        it('should match snapshot', () => {
            const document = fixtureDocument('textual_changes/before.json');
            expect(tokenize(document)).toMatchSnapshot();
        });

        it('should match snapshot', () => {
            const document = fixtureDocument('textual_changes/after.json');
            expect(tokenize(document)).toMatchSnapshot();
        });
    });
});
