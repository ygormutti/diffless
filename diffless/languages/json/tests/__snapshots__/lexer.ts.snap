// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer tokenize should match snapshot 1`] = `
Array [
  Token {
    "content": "{",
    "location": [1:1; 1:2) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 0,
  },
  ValuedToken {
    "content": "\\"errors\\"",
    "location": [2:5; 2:13) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "errors",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [2:13; 2:14) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  Token {
    "content": "[",
    "location": [2:15; 2:16) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 4,
  },
  Token {
    "content": "{",
    "location": [3:9; 3:10) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 0,
  },
  ValuedToken {
    "content": "\\"resource\\"",
    "location": [4:13; 4:23) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "resource",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [4:23; 4:24) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"Issue\\"",
    "location": [4:25; 4:32) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "Issue",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [4:32; 4:33) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"fields\\"",
    "location": [5:13; 5:21) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "fields",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [5:21; 5:22) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  Token {
    "content": "[",
    "location": [5:23; 5:24) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 4,
  },
  ValuedToken {
    "content": "\\"title\\"",
    "location": [6:17; 6:24) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "title",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [6:24; 6:25) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"author\\"",
    "location": [7:17; 7:25) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "author",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [7:25; 7:26) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"description\\"",
    "location": [8:17; 8:30) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "description",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [8:30; 8:31) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"field_with_a_really_long_name_even_longer_than_the_others_combined\\"",
    "location": [9:17; 9:85) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "field_with_a_really_long_name_even_longer_than_the_others_combined",
    "valueEquals": [Function],
  },
  Token {
    "content": "]",
    "location": [10:13; 10:14) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 5,
  },
  Token {
    "content": ",",
    "location": [10:14; 10:15) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"code\\"",
    "location": [11:13; 11:19) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "code",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [11:19; 11:20) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"missing_fields\\"",
    "location": [11:21; 11:37) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "missing_fields",
    "valueEquals": [Function],
  },
  Token {
    "content": "}",
    "location": [12:9; 12:10) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 1,
  },
  Token {
    "content": ",",
    "location": [12:10; 12:11) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  Token {
    "content": "{",
    "location": [13:9; 13:10) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 0,
  },
  ValuedToken {
    "content": "\\"resource\\"",
    "location": [14:13; 14:23) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "resource",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [14:23; 14:24) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"Issue\\"",
    "location": [14:25; 14:32) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "Issue",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [14:32; 14:33) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"field\\"",
    "location": [15:13; 15:20) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "field",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [15:20; 15:21) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"milestone\\"",
    "location": [15:22; 15:33) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "milestone",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [15:33; 15:34) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"code\\"",
    "location": [16:13; 16:19) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "code",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [16:19; 16:20) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"invalid\\"",
    "location": [16:21; 16:30) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "invalid",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [16:30; 16:31) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"value\\"",
    "location": [17:13; 17:20) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "value",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [17:20; 17:21) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  ValuedToken {
    "content": "1.0",
    "location": [17:22; 17:25) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 7,
    "value": 1,
    "valueEquals": [Function],
  },
  Token {
    "content": "}",
    "location": [18:9; 18:10) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 1,
  },
  Token {
    "content": "]",
    "location": [19:5; 19:6) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 5,
  },
  Token {
    "content": ",",
    "location": [19:6; 19:7) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"message\\"",
    "location": [20:5; 20:14) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "message",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [20:14; 20:15) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"Validation Failed\\"",
    "location": [21:9; 21:28) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 6,
    "value": "Validation Failed",
    "valueEquals": [Function],
  },
  Token {
    "content": "}",
    "location": [22:1; 22:2) @ file://diffless/tests/fixtures/textual_changes/before.json,
    "type": 1,
  },
]
`;

exports[`lexer tokenize should match snapshot 2`] = `
Array [
  Token {
    "content": "{",
    "location": [1:1; 1:2) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 0,
  },
  ValuedToken {
    "content": "\\"errors\\"",
    "location": [2:3; 2:11) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "errors",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [2:11; 2:12) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  Token {
    "content": "[",
    "location": [2:13; 2:14) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 4,
  },
  Token {
    "content": "{",
    "location": [3:5; 3:6) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 0,
  },
  ValuedToken {
    "content": "\\"resource\\"",
    "location": [4:7; 4:17) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "resource",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [4:17; 4:18) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"Issue\\"",
    "location": [4:19; 4:26) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "Issue",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [4:26; 4:27) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"fields\\"",
    "location": [5:7; 5:15) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "fields",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [5:15; 5:16) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  Token {
    "content": "[",
    "location": [5:17; 5:18) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 4,
  },
  ValuedToken {
    "content": "\\"field_with_a_really_long_name_even_longer_than_the_others_combined\\"",
    "location": [6:11; 6:79) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "field_with_a_really_long_name_even_longer_than_the_others_combined",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [6:79; 6:80) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"title\\"",
    "location": [7:11; 7:18) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "title",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [7:18; 7:19) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"description\\"",
    "location": [8:11; 8:24) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "description",
    "valueEquals": [Function],
  },
  Token {
    "content": "]",
    "location": [9:9; 9:10) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 5,
  },
  Token {
    "content": ",",
    "location": [9:10; 9:11) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"code\\"",
    "location": [10:7; 10:13) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "code",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [10:13; 10:14) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"missing_fields\\"",
    "location": [10:15; 10:31) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "missing_fields",
    "valueEquals": [Function],
  },
  Token {
    "content": "}",
    "location": [11:5; 11:6) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 1,
  },
  Token {
    "content": ",",
    "location": [11:6; 11:7) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  Token {
    "content": "{",
    "location": [12:5; 12:6) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 0,
  },
  ValuedToken {
    "content": "\\"resource\\"",
    "location": [13:7; 13:17) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "resource",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [13:17; 13:18) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"Issue\\"",
    "location": [13:19; 13:26) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "Issue",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [13:26; 13:27) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"field\\"",
    "location": [14:7; 14:14) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "field",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [14:14; 14:15) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"milestone\\"",
    "location": [14:16; 14:27) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "milestone",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [14:27; 14:28) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"code\\"",
    "location": [15:7; 15:13) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "code",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [15:13; 15:14) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"invalid\\"",
    "location": [15:15; 15:24) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "invalid",
    "valueEquals": [Function],
  },
  Token {
    "content": ",",
    "location": [15:24; 15:25) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"value\\"",
    "location": [16:7; 16:14) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "value",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [16:14; 16:15) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  ValuedToken {
    "content": "1",
    "location": [16:16; 16:17) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 7,
    "value": 1,
    "valueEquals": [Function],
  },
  Token {
    "content": "}",
    "location": [17:5; 17:6) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 1,
  },
  Token {
    "content": "]",
    "location": [18:3; 18:4) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 5,
  },
  Token {
    "content": ",",
    "location": [18:4; 18:5) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 3,
  },
  ValuedToken {
    "content": "\\"message\\"",
    "location": [19:3; 19:12) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "message",
    "valueEquals": [Function],
  },
  Token {
    "content": ":",
    "location": [19:12; 19:13) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 2,
  },
  ValuedToken {
    "content": "\\"Validation Failed\\"",
    "location": [19:14; 19:33) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 6,
    "value": "Validation Failed",
    "valueEquals": [Function],
  },
  Token {
    "content": "}",
    "location": [20:1; 20:2) @ file://diffless/tests/fixtures/textual_changes/after.json,
    "type": 1,
  },
]
`;
